### ISSUE TITLE
GKD Example why do not use labels?

### ISSUE THREAD

**Issue by YooSungHyun** [huggingface/trl#2732]
### Reproduction

```python
"""
# Full training:
python examples/scripts/gkd.py \
    --model_name_or_path Qwen/Qwen2-0.5B-Instruct \
    --teacher_model_name_or_path Qwen/Qwen2-1.5B-Instruct \
    --dataset_name trl-lib/chatbot_arena_completions \
    --learning_rate 2e-5 \
    --per_device_train_batch_size 4 \
    --gradient_accumulation_steps 8 \
    --output_dir gkd-model \
    --logging_steps 10 \
    --num_train_epochs 1 \
    --push_to_hub \
    --gradient_checkpointing

"""

import os
from accelerate import PartialState
from datasets import load_dataset
from transformers import AutoTokenizer, GenerationConfig
from trl import (
    GKDConfig,
    GKDTrainer,
    LogCompletionsCallback,
    ModelConfig,
    ScriptArguments,
    TrlParser,
    get_kbit_device_map,
    get_peft_config,
    get_quantization_config,
)


os.environ["TORCHDYNAMO_DISABLE"] = "1"

if __name__ == "__main__":
    parser = TrlParser((ScriptArguments, GKDConfig, ModelConfig))
    script_args, training_args, model_args = parser.parse_args_and_config()
    ################
    # Model & Tokenizer
    ################
    quantization_config = get_quantization_config(model_args)
    model_kwargs = dict(
        revision=model_args.model_revision,
        trust_remote_code=model_args.trust_remote_code,
        attn_implementation=model_args.attn_implementation,
        torch_dtype=model_args.torch_dtype,
        use_cache=False if training_args.gradient_checkpointing else True,
        device_map=get_kbit_device_map() if quantization_config is not None else None,
        quantization_config=quantization_config,
    )
    training_args.model_init_kwargs = model_kwargs

    teacher_model_kwargs = dict(
        revision=model_args.model_revision,
        trust_remote_code=model_args.trust_remote_code,
        attn_implementation=model_args.attn_implementation,
        torch_dtype=model_args.torch_dtype,
        use_cache=True,
        device_map=get_kbit_device_map() if quantization_config is not None else None,
        quantization_config=quantization_config,
    )
    training_args.teacher_model_init_kwargs = teacher_model_kwargs

    tokenizer = AutoTokenizer.from_pretrained(
        model_args.model_name_or_path,
        revision=model_args.model_revision,
        trust_remote_code=model_args.trust_remote_code,
        padding_side="left",
    )
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token

    # ################
    # # Dataset
    # ################
    dataset = load_dataset(script_args.dataset_name, name=script_args.dataset_config)

    with PartialState().local_main_process_first():
        dataset = dataset.map(
            lambda x: {
                "prompt": tokenizer.apply_chat_template(x["prompt"], tokenize=False, add_generation_prompt=True)
            },
            num_proc=training_args.dataset_num_proc,
        )

    # ################
    # # Training
    # ################
    trainer = GKDTrainer(
        model=model_args.model_name_or_path,
        teacher_model=training_args.teacher_model_name_or_path,
        args=training_args,
        train_dataset=dataset[script_args.dataset_train_split],
        eval_dataset=dataset[script_args.dataset_test_split] if training_args.eval_strategy != "no" else None,
        processing_class=tokenizer,
        peft_config=get_peft_config(model_args),
    )

    if training_args.eval_strategy != "no":
        generation_config = GenerationConfig(
            max_new_tokens=training_args.max_new_tokens, do_sample=True, temperature=training_args.temperature
        )
        completions_callback = LogCompletionsCallback(trainer, generation_config, num_prompts=8)
        trainer.add_callback(completions_callback)

    trainer.train()

    # Save and push to hub
    trainer.save_model(training_args.output_dir)
    if training_args.push_to_hub:
        trainer.push_to_hub(dataset_name=script_args.dataset_name)
```

If we use the `trl-lib/chatbot_arena_completions` dataset and the `prompt` column, there will be no corresponding assistant data to serve as labels.

As far as I know, labels are not required only in the following cases:
1. When using only the on-policy loss (i.e., `lambda` is `1`).
2. When `seq_kd` is set to `True` (since the teacher's output is used as supervised labels).

However, when checking the arguments of the `GKDConfig` class, I see that `lambda` is set to `0.5` and `seq_kd` is `False`. Therefore, I believe that labels are necessary for this training to work correctly.

What is the reason for structuring the examples in this way? Or am I misunderstanding GKD, such that labels are not actually required? Alternatively, is the code implemented in a way that, when labels are missing, it ensures that case 1 or 2 is applied so that training remains unaffected?

### System Info

python==3.11.11
torch==2.5.1
transformers==4.48.2
deepspeed==0.16.3
trl==0.13.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by YooSungHyun** [comment]
I think that during the `super.__init__` stage, the ChatML format is recognized, and `formatting_func` is defined to process `messages` into `input_ids`, which are then passed back to `GKDTrainer`.  
After that, it seems to be handled by `DataCollatorForChatML`.

However, ChatML does not support multi-turn processing, does it?  
In the case of multi-turn conversations, all `User` messages should be assigned `-100`, and only `Assistant` messages should be used for loss computation, right?

**Comment by albertvillanova** [comment]
Hey @YooSungHyun, sorry for the very late reply — we're catching up on the backlog. 

Your premise is right: with `lmbda=0.5` and `seq_kd=False` you do need labels, and the good news is they *are* there. `trl-lib/chatbot_arena_completions` isn't prompt-only — it has a `messages` column with the full conversation (user + assistant turns). `DataCollatorForChatML` tokenizes the whole conversation, masks everything up to the last message with -100, and keeps the final assistant turn's tokens as the labels (see https://github.com/huggingface/trl/blob/f5631147b47e9105312887ff813e4e9e28b22d2d/trl/experimental/utils.py#L389-L392). The `prompt` mapping in the old example script only prepared the generation prompt for on-policy sampling; it doesn't strip the labels. On multi-turn: you're correct that only the final assistant turn contributes to the loss — earlier turns (including earlier assistant replies) are treated as prompt context, which is intentional for GKD's prompt/completion split. 

Closing as answered, but happy to re-open if anything's still unclear. (Note: GKD now lives under `trl.experimental.gkd`.)
